src/linter.cc

```diff
@@ -22,92 +22,245 @@
+            .end().GetByteOffset();
+
+        if ( location >= sql.size() || sql[location] != ';' ) {
+            return absl::Status(
+                    absl::StatusCode::kFailedPrecondition,
+                    "Each statemnt should end with a consequtive"
+                    "semicolon ';'");
+        }
+    }
+    return absl::OkStatus();
+}
+
+bool AllUpperCase(const absl::string_view &sql,
+    const ParseLocationRange &range) {
+    for (int i = range.start().GetByteOffset();
+                i < range.end().GetByteOffset(); ++i) {
+        if ( 'a' <= sql[i] && sql[i] <= 'z' )
+            return false;
+    }
     return true;
 }
 
-absl::Status checkUppercaseKeywords(absl::string_view sql) {
-    std::vector<const zetasql::ASTNode *> keywords{};
+absl::Status CheckUppercaseKeywords(absl::string_view sql) {
+    ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
+    std::vector<ParseToken> parse_tokens;
+    absl::Status tokenizer_status =
+        GetParseTokens(ParseTokenOptions(), &location, &parse_tokens);
+
+    // Keyword definition in tokenizer is very wide,
+    // it include some special characters like ';', '*', etc.
+    // Keyword Uppercase check will simply ignore characters
+    // outside of english lowercase letters.
+    for ( auto &token : parse_tokens ) {
+      if ( token.kind() == ParseToken::KEYWORD ) {
+        if (!AllUpperCase(sql, token.GetLocationRange())) {
+          return absl::Status(
+            absl::StatusCode::kFailedPrecondition,
+            absl::StrCat("All keywords should be Uppercase, In character ",
+            std::to_string(token.GetLocationRange().start().GetByteOffset()),
+            " string should be: ",
+            token.GetSQL()));
+        }
+      }
+    }
+
+    return absl::OkStatus();
+}
+
+absl::Status CheckCommentType(absl::string_view sql) {
+    bool includes_type1 = false;
+    bool includes_type2 = false;
+    bool includes_type3 = false;
+    bool inside_string = false;
 
-    std::unique_ptr<zetasql::ParserOutput> output;
+    for (int i = 1; i<static_cast<int>(sql.size()); ++i) {
+        if (!inside_string && sql[i-1] == '-' && sql[i] == '-') {
+            includes_type1 = true;
+            // ignore the line.
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    absl::Status parser_status = zetasql::ParseStatement(sql,
-            zetasql::ParserOptions(), &output);
+        if (!inside_string && sql[i-1] == '/' && sql[i] == '/') {
+            includes_type2 = true;
+            // ignore the line.
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    if ( !parser_status.ok() )
-        return parser_status;
-    // TODO(orhan_uysal): Implement KeywordExtractor.
+        if (!inside_string && sql[i] == '#') {
```

Comment:
and I guess, you also have a bug if there are no characters after a comment:
`-- Comment` (no new line at the end)
`/* Unfinished comment`

Reply:
I think it should be ok. If there is no new line at the end that means they are on the same line. Then '/*' will be just seen as a comment, not a multiline comment starter. Am I missing something in there?